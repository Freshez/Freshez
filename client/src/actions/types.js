export const SET_ALERT = 'SET_ALERT';
export const REMOVE_ALERT = 'REMOVE_ALERT';
export const REGISTER_SUCCESS = 'REGISTER_SUCCESS';
export const REGISTER_FAIL = 'REGISTER_FAIL';
export const USER_LOADED = 'USER_LOADED';
export const AUTH_ERROR = 'AUTH_ERROR';
export const LOGIN_SUCCESS = 'LOGIN_SUCCESS';
export const LOGIN_FAIL = 'LOGIN_FAIL';
export const LOGOUT = 'LOGOUT';
export const GET_PROFILE = 'GET_PROFILE';
export const GET_PROFILES = 'GET_PROFILES';
export const CLEAR_PROFILE = 'CLEAR_PROFILE';
export const PROFILE_ERROR = 'PROFILE_ERROR';
export const ACCOUNT_DELETED = 'ACCOUNT_DELETED';
export const GET_POSTS = 'GET_POSTS';
export const GET_POST = 'GET_POST';
export const POST_ERROR = 'POST_ERROR';
export const UPDATE_LIKES = 'UPDATE_LIKES';
export const DELETE_POST = 'DELETE_POST';
export const ADD_POST = 'ADD_POST';
export const SET_LOADING_TRUE = 'SET_LOADING_TRUE';
export const SET_LOADING_FALSE = 'SET_LOADING_FALSE';
export const ADD_COMMENT = 'ADD_COMMENT';
export const REMOVE_COMMENT = 'REMOVE_COMMENT';
export const LOGIN_OPEN = 'LOGIN_OPEN';
export const LOGIN_CLOSED = 'LOGIN_CLOSED';
export const REGISTER_OPEN = 'REGISTER_OPEN';
export const REGISTER_CLOSED = 'REGISTER_CLOSED';
export const CREATE_VENDOR_PROFILE_OPEN = 'CREATE_VENDOR_PROFILE_OPEN';
export const CREATE_VENDOR_PROFILE_CLOSED = 'CREATE_VENDOR_PROFILE_CLOSED';
export const CREATE_CUSTOMER_PROFILE_OPEN = 'CREATE_CUSTOMER_PROFILE_OPEN';
export const CREATE_CUSTOMER_PROFILE_CLOSED = 'CREATE_CUSTOMER_PROFILE_CLOSED';
export const REGISTER_BUTTON_VENDOR = 'REGISTER_BUTTON_VENDOR';
export const REGISTER_BUTTON_CUSTOMER = 'REGISTER_BUTTON_CUSTOMER';
export const DESELECT_ROLE_BUTTON = 'DESELECT_ROLE_BUTTON';
export const CLEAR_VENDOR_POSTS = 'CLEAR_VENDOR_POSTS';
export const GET_VENDOR_POSTS = 'GET_VENDOR_POSTS';
