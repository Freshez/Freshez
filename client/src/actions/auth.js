import axios from 'axios';
import { setAlert } from './alert';
import {
  REGISTER_SUCCESS,
  REGISTER_FAIL,
  USER_LOADED,
  AUTH_ERROR,
  LOGIN_SUCCESS,
  LOGIN_FAIL,
  LOGOUT,
  CLEAR_PROFILE,
  REGISTER_BUTTON_VENDOR,
  REGISTER_BUTTON_CUSTOMER,
  DESELECT_ROLE_BUTTON,
} from './types';
import setAuthToken from '../utils/setAuthToken';

// Load User
export const loadUser = () => async (dispatch) => {
  if (localStorage.token) {
    setAuthToken(localStorage.token);
  }

  try {
    const res = await axios.get('/api/auth');

    dispatch({
      type: USER_LOADED,
      payload: res.data,
    });
  } catch (err) {
    dispatch({
      type: AUTH_ERROR,
    });
  }
};

// Register User
export const register = ({ name, role, email, password }) => async (
  dispatch
) => {
  const config = {
    headers: {
      'Content-Type': 'application/json',
    },
  };

  const body = JSON.stringify({ name, role, email, password });

  try {
    const res = await axios.post('/api/users', body, config);

    dispatch({
      type: REGISTER_SUCCESS,
      payload: res.data,
    });

    dispatch(loadUser());
  } catch (err) {
    const errors = err.response.data.errors;

    if (errors) {
      errors.forEach((error) => dispatch(setAlert(error.msg, 'danger')));
    }
    dispatch({
      type: REGISTER_FAIL,
    });
  }
};

// Login User
export const login = (email, password) => async (dispatch) => {
  const config = {
    headers: {
      'Content-Type': 'application/json',
    },
  };

  const body = JSON.stringify({ email, password });

  try {
    const res = await axios.post('/api/auth', body, config);

    dispatch({
      type: LOGIN_SUCCESS,
      payload: res.data,
    });

    dispatch(loadUser());
  } catch (err) {
    const errors = err.response.data.errors;

    if (errors) {
      errors.forEach((error) => dispatch(setAlert(error.msg, 'danger')));
    }
    dispatch({
      type: LOGIN_FAIL,
    });
  }
};

// Logout / Clear Profile
export const logout = () => (dispatch) => {
  dispatch({ type: CLEAR_PROFILE });
  dispatch({ type: LOGOUT });
};

// Click Vendor Button
export const selectVendor = () => (dispatch) => {
  dispatch({ type: REGISTER_BUTTON_VENDOR });
};

// Click Customer Button
export const selectCustomer = () => (dispatch) => {
  dispatch({ type: REGISTER_BUTTON_CUSTOMER });
};

export const deselectRole = () => (dispatch) => {
  dispatch({ type: DESELECT_ROLE_BUTTON });
};
