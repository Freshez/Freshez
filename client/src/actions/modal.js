import axios from 'axios';
import { setAlert } from './alert';
import {
  LOGIN_OPEN,
  LOGIN_CLOSED,
  REGISTER_OPEN,
  REGISTER_CLOSED,
  CREATE_VENDOR_PROFILE_OPEN,
  CREATE_VENDOR_PROFILE_CLOSED,
  CREATE_CUSTOMER_PROFILE_OPEN,
  CREATE_CUSTOMER_PROFILE_CLOSED,
} from './types';

// Open login modal
export const openLoginModal = () => async (dispatch) => {
  try {
    dispatch({
      type: LOGIN_OPEN,
    });
  } catch (err) {}
};

// Close login modal
export const closeLoginModal = () => async (dispatch) => {
  try {
    dispatch({
      type: LOGIN_CLOSED,
    });
  } catch (err) {}
};

// Open login modal
export const openRegisterModal = () => async (dispatch) => {
  try {
    dispatch({
      type: REGISTER_OPEN,
    });
  } catch (err) {}
};

// Close login modal
export const closeRegisterModal = () => async (dispatch) => {
  try {
    dispatch({
      type: REGISTER_CLOSED,
    });
  } catch (err) {}
};

// Open create vendor profile modal
export const openCreateVendorProfileModal = () => async (dispatch) => {
  try {
    dispatch({
      type: CREATE_VENDOR_PROFILE_OPEN,
    });

    console.log('yes');
  } catch (err) {
    console.log('no');
  }
};

// Close create vendor profile modal
export const closeCreateVendorProfileModal = () => async (dispatch) => {
  try {
    dispatch({
      type: CREATE_VENDOR_PROFILE_CLOSED,
    });
  } catch (err) {}
};

// Open create customer profile modal
export const openCreateCustomerProfileModal = () => async (dispatch) => {
  try {
    dispatch({
      type: CREATE_CUSTOMER_PROFILE_OPEN,
    });
  } catch (err) {}
};

// Close create customer profile modal
export const closeCreateCustomerProfileModal = () => async (dispatch) => {
  try {
    dispatch({
      type: CREATE_CUSTOMER_PROFILE_CLOSED,
    });
  } catch (err) {}
};
