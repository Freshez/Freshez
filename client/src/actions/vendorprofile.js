import axios from 'axios';
import { setAlert } from './alert';

import {
  GET_PROFILE,
  GET_PROFILES,
  PROFILE_ERROR,
  CLEAR_PROFILE,
  ACCOUNT_DELETED,
} from './types';

// Get current users profile
export const getCurrentProfile = () => async (dispatch) => {
  try {
    const res = await axios.get('/api/vendorprofile/me');

    dispatch({
      type: GET_PROFILE,
      payload: res.data,
    });
  } catch (err) {
    dispatch({
      type: PROFILE_ERROR,
      payload: {
        msg: err.response.statusText,
        status: err.response.status,
      },
    });
  }
};

// Get vendor profile by id
export const getProfileById = (userId) => async (dispatch) => {
  try {
    const res = await axios.get(`/api/vendorprofile/user/${userId}`);

    dispatch({
      type: GET_PROFILE,
      payload: res.data,
    });
  } catch (err) {
    dispatch({
      type: PROFILE_ERROR,
      payload: {
        msg: err.response.statusText,
        status: err.response.status,
      },
    });
  }
};

// Get all vendor profiles
export const getProfiles = () => async (dispatch) => {
  dispatch({ type: CLEAR_PROFILE });
  try {
    const res = await axios.get('/api/vendorprofile');

    dispatch({
      type: GET_PROFILES,
      payload: res.data,
    });
  } catch (err) {
    dispatch({
      type: PROFILE_ERROR,
      payload: {
        msg: err.response.statusText,
        status: err.response.status,
      },
    });
  }
};

// Create or update profile
export const createVendorProfile = (formData, history, edit = false) => async (
  dispatch
) => {
  try {
    const config = {
      headers: {
        'Content-Type': 'application/json',
      },
    };

    var pattern_no_protocol = new RegExp(
      '^((([a-z\\d]([a-z\\d-]*[a-z\\d])*)\\.)+[a-z]{2,}|' + // domain name
      '((\\d{1,3}\\.){3}\\d{1,3}))' + // OR ip (v4) address
      '(\\:\\d+)?(\\/[-a-z\\d%_.~+]*)*' + // port and path
      '(\\?[;&a-z\\d%_.~+=-]*)?' + // query string
        '(\\#[-a-z\\d_]*)?$',
      'i'
    ); // fragment locator

    var pattern_required_protocol = new RegExp(
      '^(https?:\\/\\/)+' + // protocol
      '((([a-z\\d]([a-z\\d-]*[a-z\\d])*)\\.)+[a-z]{2,}|' + // domain name
      '((\\d{1,3}\\.){3}\\d{1,3}))' + // OR ip (v4) address
      '(\\:\\d+)?(\\/[-a-z\\d%_.~+]*)*' + // port and path
      '(\\?[;&a-z\\d%_.~+=-]*)?' + // query string
        '(\\#[-a-z\\d_]*)?$',
      'i'
    ); // fragment locator

    console.log(formData.website);
    if (!!pattern_no_protocol.test(formData.website)) {
      console.log('match');
      formData.website = 'http://' + formData.website;
    }

    if (!pattern_required_protocol.test(formData.website)) {
      return dispatch(setAlert('Invalid Website URL', 'danger'));
    }

    const res = await axios.post('/api/vendorprofile', formData, config);

    dispatch({
      type: GET_PROFILE,
      payload: res.data,
    });

    dispatch(setAlert(edit ? 'Profile Updated' : 'Profile Created'));

    if (!edit) {
      history.push('/dashboard');
    }
  } catch (err) {
    const errors = err.response.data.errors;

    if (errors) {
      errors.forEach((error) => dispatch(setAlert(error.msg, 'danger')));
    }

    dispatch({
      type: PROFILE_ERROR,
      payload: {
        msg: err.response.statusText,
        status: err.response.status,
      },
    });
  }
};

// Delete account & profile
export const deleteAccount = (id) => async (dispatch) => {
  if (window.confirm('Are you sure? This can NOT be undone')) {
    try {
      await axios.delete('/api/profile');

      dispatch({
        type: CLEAR_PROFILE,
      });

      dispatch({
        type: ACCOUNT_DELETED,
      });

      dispatch(setAlert('Your account has been permanently deleted'));
    } catch (err) {
      dispatch({
        type: PROFILE_ERROR,
        payload: { msg: err.response.statusText, status: err.response.status },
      });
    }
  }
};
