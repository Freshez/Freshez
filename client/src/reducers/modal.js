import {
  LOGIN_OPEN,
  LOGIN_CLOSED,
  REGISTER_OPEN,
  REGISTER_CLOSED,
  CREATE_VENDOR_PROFILE_OPEN,
  CREATE_VENDOR_PROFILE_CLOSED,
  CREATE_CUSTOMER_PROFILE_OPEN,
  CREATE_CUSTOMER_PROFILE_CLOSED,
} from '../actions/types';

const initialState = {
  login_open: false,
  register_open: false,
  create_vendor_open: false,
  create_customer_open: false,
};

export default function (state = initialState, action) {
  const { type, payload } = action;

  switch (type) {
    case LOGIN_OPEN:
      return {
        ...state,
        login_open: true,
      };
    case LOGIN_CLOSED:
      return {
        ...state,
        login_open: false,
      };
    case REGISTER_OPEN:
      return {
        ...state,
        register_open: true,
      };
    case REGISTER_CLOSED:
      return {
        ...state,
        register_open: false,
      };
    case CREATE_VENDOR_PROFILE_OPEN:
      localStorage.setItem('createvendorprofile', true);
      return {
        ...state,
        create_vendor_open: true,
      };
    case CREATE_VENDOR_PROFILE_CLOSED:
      localStorage.setItem('createvendorprofile', false);
      return {
        ...state,
        create_vendor_open: false,
      };
    case CREATE_CUSTOMER_PROFILE_OPEN:
      localStorage.setItem('createcustomerprofile', true);
      return {
        ...state,
        create_customer_open: true,
      };
    case CREATE_CUSTOMER_PROFILE_CLOSED:
      localStorage.setItem('createcustomerprofile', true);
      return {
        ...state,
        create_customer_open: false,
      };
    default:
      return state;
  }
}
