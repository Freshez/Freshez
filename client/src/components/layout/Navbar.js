import React, { Fragment, useState } from 'react';
import { Link } from 'react-router-dom';
import { connect } from 'react-redux';
import PropTypes from 'prop-types';
import { logout } from '../../actions/auth';
import {
  openLoginModal,
  closeLoginModal,
  openRegisterModal,
  closeRegisterModal,
} from '../../actions/modal';

const Navbar = ({
  auth: { isAuthenticated, loading },
  logout,
  openLoginModal,
  openRegisterModal,
  closeLoginModal,
  closeRegisterModal,
  loginModalOpen,
  regsiterModalOpen,
}) => {
  const [hovered_login, setHoveredLogin] = useState(false);
  const [hovered_dev_connector, setHoveredDevConnector] = useState(false);
  const [hovered_register, setHoveredRegister] = useState(false);
  const [hovered_vendors, setHoveredVendors] = useState(false);
  const [hovered_posts, setHoveredPosts] = useState(false);
  const [hovered_dashboard, setHoveredDashboard] = useState(false);
  const [hovered_logout, setHoveredLogout] = useState(false);
  const toggleHoverLogin = () => setHoveredLogin(!hovered_login);
  const toggleHoverDevConnector = () =>
    setHoveredDevConnector(!hovered_dev_connector);
  const toggleHoverRegister = () => setHoveredRegister(!hovered_register);
  const toggleHoverVendors = () => setHoveredVendors(!hovered_vendors);
  const toggleHoverPosts = () => setHoveredPosts(!hovered_posts);
  const toggleHoverDashboard = () => setHoveredDashboard(!hovered_dashboard);
  const toggleHoverLogout = () => setHoveredLogout(!hovered_logout);
  const authLinks = (
    <ul>
      <li>
        <div
          className={hovered_vendors ? 'nav-link-hovered nav-link' : 'nav-link'}
          onMouseEnter={toggleHoverVendors}
          onMouseLeave={toggleHoverVendors}
        >
          <Link to='/vendorprofiles'>Vendors</Link>
        </div>
      </li>
      <li>
        <div
          className={hovered_posts ? 'nav-link-hovered nav-link' : 'nav-link'}
          onMouseEnter={toggleHoverPosts}
          onMouseLeave={toggleHoverPosts}
        >
          <Link to='/posts'>Posts</Link>
        </div>
      </li>
      <li>
        <div
          className={
            hovered_dashboard ? 'nav-link-hovered nav-link' : 'nav-link'
          }
          onMouseEnter={toggleHoverDashboard}
          onMouseLeave={toggleHoverDashboard}
        >
          <Link to='/dashboard'>
            <i className='fas fa-user'></i>{' '}
            <span className='hide-sm'>Dashboard</span>
          </Link>
        </div>
      </li>
      <li>
        <div
          className={hovered_logout ? 'nav-link-hovered nav-link' : 'nav-link'}
          onMouseEnter={toggleHoverLogout}
          onMouseLeave={toggleHoverLogout}
        >
          <a onClick={logout} href='#!'>
            <i className='fas fa-sign-out-alt'></i>{' '}
            <span className='hide-sm'>Logout</span>
          </a>
        </div>
      </li>
    </ul>
  );

  const guestLinks = (
    <ul>
      <li>
        <div
          className={hovered_vendors ? 'nav-link-hovered nav-link' : 'nav-link'}
          onMouseEnter={toggleHoverVendors}
          onMouseLeave={toggleHoverVendors}
        >
          <Link to='/vendorprofiles'>Vendors</Link>
        </div>
      </li>
      <li>
        <div
          className={
            hovered_register ? 'nav-link-hovered nav-link' : 'nav-link'
          }
          onMouseEnter={toggleHoverRegister}
          onMouseLeave={toggleHoverRegister}
        >
          <Link id='registerLink' to='' onClick={() => openRegisterModal()}>
            Register
          </Link>
        </div>
      </li>
      <li>
        <div
          className={hovered_login ? 'nav-link-hovered nav-link' : 'nav-link'}
          onMouseEnter={toggleHoverLogin}
          onMouseLeave={toggleHoverLogin}
        >
          <Link id='loginLink' to='' onClick={() => openLoginModal()}>
            Login
          </Link>
        </div>
      </li>
    </ul>
  );

  return (
    <nav className='navbar bg-white'>
      <h1>
        <div
          className={
            hovered_dev_connector ? 'nav-link-hovered nav-link' : 'nav-link'
          }
          onMouseEnter={toggleHoverDevConnector}
          onMouseLeave={toggleHoverDevConnector}
        >
          <Link to='/'>
            <i className='fas fa-code'></i> Freshez
          </Link>
        </div>
      </h1>
      {!loading && (
        <Fragment>{isAuthenticated ? authLinks : guestLinks}</Fragment>
      )}
    </nav>
  );
};

Navbar.propTypes = {
  logout: PropTypes.func.isRequired,
  auth: PropTypes.object.isRequired,
  loginModalOpen: PropTypes.bool.isRequired,
  regsiterModalOpen: PropTypes.bool.isRequired,
};

const mapStateToProps = (state) => ({
  auth: state.auth,
  vendorprofile: state.vendorprofile,
  customerprofile: state.customerprofile,
  loginModalOpen: state.modal.login_open,
  regsiterModalOpen: state.modal.register_open,
});

export default connect(mapStateToProps, {
  logout,
  openLoginModal,
  openRegisterModal,
  closeLoginModal,
  closeRegisterModal,
})(Navbar);
