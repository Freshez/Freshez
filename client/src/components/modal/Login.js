import React, { Fragment, useState } from 'react';
import { connect } from 'react-redux';
import { Link, Redirect } from 'react-router-dom';
import PropTypes from 'prop-types';
import ReactDOM from 'react-dom';
import Modal from 'react-modal';
import { login } from '../../actions/auth';

import {
  openLoginModal,
  closeLoginModal,
  openRegisterModal,
  closeRegisterModal,
} from '../../actions/modal';

const customStyles = {
  content: {
    top: '50%',
    left: '50%',
    right: 'auto',
    bottom: 'auto',
    marginRight: '-50%',
    transform: 'translate(-50%, -50%)',
    height: '75%',
  },
};

const Login = ({
  isAuthenticated,
  openLoginModal,
  openRegisterModal,
  closeLoginModal,
  closeRegisterModal,
  loginModalOpen,
  regsiterModalOpen,
  login,
}) => {
  var subtitle;
  const [formData, setFormData] = useState({
    email: '',
    password: '',
  });

  const { email, password } = formData;

  const onSubmit = async (e) => {
    e.preventDefault();
    login(email, password);
  };

  // Set form data to copy of formData with changed value
  const onChange = (e) =>
    setFormData({ ...formData, [e.target.name]: e.target.value });

  const [modalIsOpen, setIsOpen] = React.useState(false);
  function openModal() {
    setIsOpen(true);
    openLoginModal();
  }

  function closeModal() {
    setIsOpen(false);
    closeLoginModal();
  }

  const [show, setShow] = useState(false);

  const handleClose = () => setShow(false);
  const handleShow = () => setShow(true);

  // Redirect if logged in TODO - close modal if authenticated?
  if (isAuthenticated) {
    return <Redirect to='/dashboard' />;
  }
  return (
    <div>
      <Modal
        isOpen={loginModalOpen}
        onRequestClose={closeModal}
        style={customStyles}
        contentLabel='Example Modal'
      >
        <Fragment>
          <div className='close-btn-outline'>
            <button
              type='button'
              className='btn close-btn'
              aria-label='Close'
              onClick={() => closeLoginModal()}
            >
              <span aria-hidden='true'>&times;</span>
            </button>
          </div>
          <h1 className='large text-info'>Sign In</h1>
          <p className='lead'>
            <i className='fas fa-user'></i> Sign Into Your Account
          </p>
          <form className='form' onSubmit={(e) => onSubmit(e)}>
            <div className='form-group'>
              <input
                type='email'
                placeholder='Email Address'
                value={email}
                onChange={(e) => onChange(e)}
                name='email'
                required
              />
            </div>
            <div className='form-group'>
              <input
                type='password'
                placeholder='Password'
                name='password'
                minLength='6'
                value={password}
                onChange={(e) => onChange(e)}
              />
            </div>
            <input type='submit' className='btn btn-primary' value='Login' />
          </form>
          <p className='my-1'>
            Don't have an account? <Link to='/register'>Sign Up</Link>
          </p>
        </Fragment>
      </Modal>
    </div>
  );
};

Login.propTypes = {
  isAuthenticated: PropTypes.bool,
  login: PropTypes.func.isRequired,
  loginModalOpen: PropTypes.bool,
  regsiterModalOpen: PropTypes.bool,
};

const mapStateToProps = (state) => ({
  isAuthenticated: state.auth.isAuthenticated,
  loginModalOpen: state.modal.login_open,
  regsiterModalOpen: state.modal.login_open,
});

export default connect(mapStateToProps, {
  openLoginModal,
  openRegisterModal,
  closeLoginModal,
  closeRegisterModal,
  login,
})(Login);
