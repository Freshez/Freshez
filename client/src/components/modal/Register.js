import React, { Fragment, useState } from 'react';
import { connect } from 'react-redux';
import { Link, Redirect } from 'react-router-dom';
import PropTypes from 'prop-types';
import ReactDOM from 'react-dom';
import Modal from 'react-modal';
import { setAlert } from '../../actions/alert';
import {
  register,
  deselectRole,
  selectVendor,
  selectCustomer,
} from '../../actions/auth';
import {
  openLoginModal,
  closeLoginModal,
  openRegisterModal,
  closeRegisterModal,
  openCreateVendorProfileModal,
} from '../../actions/modal';
import useWindowDimensions from '../../utils/getWindowDimensions';

const customStyles = {
  content: {
    top: '55%',
    left: '50%',
    right: 'auto',
    bottom: 'auto',
    marginRight: '-50%',
    transform: 'translate(-50%, -50%)',
    height: '80%',
    width: '50%',
  },
};

const Register = ({
  isAuthenticated,
  setAlert,
  register,
  openLoginModal,
  openRegisterModal,
  closeLoginModal,
  closeRegisterModal,
  loginModalOpen,
  regsiterModalOpen,
  createVendorProfileModalOpen,
  openCreateVendorProfileModal,
  selectVendor,
  selectCustomer,
  deselectRole,
  vendorSelected,
  customerSelected,
}) => {
  var subtitle;
  const [formData, setFormData] = useState({
    name: '',
    role: '',
    email: '',
    password: '',
    password2: '',
  });

  const { height, width } = useWindowDimensions();

  if (width > 700) {
    customStyles.content = {
      top: '55%',
      left: '50%',
      right: 'auto',
      bottom: 'auto',
      marginRight: '-50%',
      transform: 'translate(-50%, -50%)',
      height: '80%',
      width: '50%',
    };
  } else {
    customStyles.content = {
      top: '60%',
      left: '50%',
      right: 'auto',
      bottom: '50%',
      marginRight: '-50%',
      transform: 'translate(-50%, -50%)',
      height: '80%',
      width: '60%',
    };
  }
  // eslint-disable-next-line
  const { name, role, email, password, password2 } = formData;

  const onSubmit = async (e) => {
    e.preventDefault();
    if (password !== password2) {
      setAlert('Passwords do not match', 'danger');
    } else {
      register({ name, role, email, password });
    }
    closeRegisterModal();
    openCreateVendorProfileModal();
    deselectRole();
  };

  // Set form data to copy of formData with changed value
  const onChange = (e) =>
    setFormData({ ...formData, [e.target.name]: e.target.value });

  // Set form data to copy of formData with changed role
  const onClick = (e) => {
    console.log('Target.name ' + e.target.name);
    console.log('Target.value ' + e.target.value);
    if (e.target.value == 'Vendor') selectVendor();
    if (e.target.value == 'Customer') selectCustomer();
    setFormData({ ...formData, [e.target.name]: e.target.value });
  };

  const openLogin = (e) => {
    closeRegisterModal();
    openLoginModal();
  };
  const [modalIsOpen, setIsOpen] = React.useState(false);
  function openModal() {
    setIsOpen(true);
    openRegisterModal();
  }

  function closeModal() {
    setIsOpen(false);
    closeRegisterModal();
  }

  // Redirect if logged in TODO - close modal if authenticated?
  if (isAuthenticated) {
    return <Redirect to='/dashboard' />;
  }
  return (
    <div>
      <Modal
        isOpen={regsiterModalOpen}
        onRequestClose={closeModal}
        style={customStyles}
        contentLabel='Example Modal'
      >
        <Fragment>
          <div className='close-btn-outline'>
            <button
              type='button'
              className='btn close-btn'
              aria-label='Close'
              onClick={() => closeRegisterModal()}
            >
              <span aria-hidden='true'>&times;</span>
            </button>
          </div>
          <h1 className='large-2 text-info'>Sign Up</h1>
          <p className='lead'>
            <i className='fas fa-user'></i> Create Your Account
          </p>
          <form className='form' onSubmit={(e) => onSubmit(e)}>
            <div className='form-group'>
              <input
                type='text'
                placeholder='Name'
                name='name'
                value={name}
                onChange={(e) => onChange(e)}
              />
            </div>
            <div className='form-group'>
              {/* <input
                type='role'
                placeholder='Customer'
                className='btn btn-role'
                name='role'
                value='Customer'
                onChange={(e) => onChange(e)}
                onClick={(e) => onClick(e)}
                // eslint-disable-next-line
                type='button'
              />
              <input
                type='role'
                placeholder='Vendor'
                className='btn btn-role'
                name='role'
                value='Vendor'
                onChange={(e) => onChange(e)}
                onClick={(e) => onClick(e)}
                // eslint-disable-next-line
                type='button'
              /> */}
              <input
                type='role'
                placeholder='Customer'
                name='role'
                value='Customer'
                onChange={(e) => onChange(e)}
                onClick={(e) => onClick(e)}
                // eslint-disable-next-line
                type='button'
                className={
                  customerSelected === true
                    ? 'btn btn-role btn-primary'
                    : 'btn btn-primary'
                }
                type='button'
              ></input>
              <input
                type='role'
                placeholder='Vendor'
                name='role'
                value='Vendor'
                onChange={(e) => onChange(e)}
                onClick={(e) => onClick(e)}
                // eslint-disable-next-line
                type='button'
                className={
                  vendorSelected === true
                    ? 'btn btn-role btn-primary'
                    : 'btn btn-primary'
                }
                type='button'
              ></input>
            </div>
            <div className='form-group'>
              <input
                type='email'
                placeholder='Email Address'
                value={email}
                onChange={(e) => onChange(e)}
                name='email'
              />
              <small className='form-text'></small>
            </div>
            <div className='form-group'>
              <input
                type='password'
                placeholder='Password'
                name='password'
                value={password}
                onChange={(e) => onChange(e)}
              />
            </div>
            <div className='form-group'>
              <input
                type='password'
                placeholder='Confirm Password'
                name='password2'
                value={password2}
                onChange={(e) => onChange(e)}
              />
            </div>
            <input type='submit' className='btn btn-primary' value='Register' />
          </form>
          <p className='my-1'>
            Already have an account?{' '}
            <Link
              to=''
              onClick={() => {
                openLogin();
              }}
            >
              Sign In
            </Link>
          </p>
        </Fragment>
      </Modal>
    </div>
  );
};

Register.propTypes = {
  openCreateVendorProfileModal: PropTypes.func.isRequired,
  isAuthenticated: PropTypes.bool,
  setAlert: PropTypes.func.isRequired,
  register: PropTypes.func.isRequired,
  loginModalOpen: PropTypes.bool,
  regsiterModalOpen: PropTypes.bool,
  createVendorProfileModalOpen: PropTypes.bool,
  vendorSelected: PropTypes.bool,
  customerSelected: PropTypes.bool,
};

const mapStateToProps = (state) => ({
  isAuthenticated: state.auth.isAuthenticated,
  loginModalOpen: state.modal.login_open,
  regsiterModalOpen: state.modal.register_open,
  createVendorProfileModalOpen: state.modal.create_vendor_open,
  vendorSelected: state.auth.vendorSelected,
  customerSelected: state.auth.customerSelected,
});

export default connect(mapStateToProps, {
  openLoginModal,
  openRegisterModal,
  closeLoginModal,
  closeRegisterModal,
  openCreateVendorProfileModal,
  register,
  setAlert,
  selectVendor,
  selectCustomer,
  deselectRole,
})(Register);
