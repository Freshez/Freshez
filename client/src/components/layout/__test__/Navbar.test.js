import React, { Fragment } from 'react';
import ReactDOM from 'react-dom';
import Navbar from '../Navbar';
import { Provider } from 'react-redux';
import { createStore, applyMiddleware } from 'redux';
import { composeWithDevTools } from 'redux-devtools-extension';
import { BrowserRouter as Router, Route, Switch } from 'react-router-dom';
import rootReducer from '../../../reducers';
import thunk from 'redux-thunk';
import renderer from 'react-test-renderer';
import { openLoginModal, openRegisterModal } from '../../../actions/modal';
import configureMockStore from 'redux-mock-store';
import Enzyme, { mount, shallow, configure } from 'enzyme';
import { Link } from 'react-router-dom';
import Adapter from 'enzyme-adapter-react-16';

configure({ adapter: new Adapter() });

const initialState = {};

const middleware = [thunk];

const store = createStore(
  rootReducer,
  initialState,
  composeWithDevTools(applyMiddleware(...middleware))
);

const mockStore = configureMockStore(middleware);

describe('My Connected React-Redux Component', () => {
  let store;
  let component;

  beforeEach(() => {
    store = mockStore({
      auth: {
        isAuthenticated: null,
        loading: false,
      },
      modal: {
        login_open: false,
        register_open: false,
      },
      logout: null,
      regsiterModalOpen: false,
    });

    store.dispatch = jest.fn();

    component = renderer.create(
      <Provider store={store}>
        <Router>
          <Fragment>
            <Navbar></Navbar>
          </Fragment>
        </Router>
      </Provider>
    );
  });

  it('should render with given state from Redux store', () => {
    expect(component.toJSON()).toMatchSnapshot();
  });

  it('should dispatch an action on button click', () => {
    renderer.act(() => {
      document.getElementById('loginLink').props.onClick();
    });
    expect(store.dispatch).toHaveBeenCalledTimes(1);
    expect(store.dispatch).toHaveBeenCalledWith(openLoginModal());
  });

  // it('should dispatch an action on button click', () => {
  //   renderer.act(() => {
  //     component.root.getElementById('registerLink').props.onClick();
  //   });
  //   expect(store.dispatch).toHaveBeenCalledTimes(1);
  //   expect(store.dispatch).toHaveBeenCalledWith(openRegisterModal());
  // });
});
