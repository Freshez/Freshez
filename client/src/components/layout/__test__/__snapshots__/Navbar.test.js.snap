// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`My Connected React-Redux Component should render with given state from Redux store 1`] = `
<nav
  className="navbar bg-white"
>
  <h1>
    <div
      className="nav-link"
      onMouseEnter={[Function]}
      onMouseLeave={[Function]}
    >
      <a
        href="/"
        onClick={[Function]}
      >
        <i
          className="fas fa-code"
        />
         Freshez
      </a>
    </div>
  </h1>
  <ul>
    <li>
      <div
        className="nav-link"
        onMouseEnter={[Function]}
        onMouseLeave={[Function]}
      >
        <a
          href="/vendorprofiles"
          onClick={[Function]}
        >
          Vendors
        </a>
      </div>
    </li>
    <li>
      <div
        className="nav-link"
        onMouseEnter={[Function]}
        onMouseLeave={[Function]}
      >
        <a
          href="/"
          id="registerLink"
          onClick={[Function]}
        >
          Register
        </a>
      </div>
    </li>
    <li>
      <div
        className="nav-link"
        onMouseEnter={[Function]}
        onMouseLeave={[Function]}
      >
        <a
          href="/"
          id="loginLink"
          onClick={[Function]}
        >
          Login
        </a>
      </div>
    </li>
  </ul>
</nav>
`;
